Add Modulemd.Defaults objects

sgallagher opened the pull request:
This implements only the abstract parent class (plus just enough of the `Modulemd.DefaultsV1` class to be able to run the tests).

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

sgallagher commented:
OK, I made the changes I mentioned in the comment above and pushed them as a fixup patch so it's easier to see the differences. If the combined set passes code-review, I'll squash them together prior to merging.

puiterwijk commented on modulemd/v2/modulemd-defaults.c:
```diff
@@ -0,0 +1,309 @@
+ModulemdDefaults *
+modulemd_defaults_new (guint64 mdversion, const gchar *module_name)
+{
+  g_return_val_if_fail (mdversion && mdversion <= MD_DEFAULTS_VERSION_LATEST,
```
Shouldn't this be `MD_DEFAULTS_VERSION_ONE`, since this would seem to imply that when a Defaultsv2 is added, it is no longer possible to create a DefaultsV1.
While I guess your intention would be to modify this later, maybe it's an idea to make this intention clear by either using `VERSION_ONE`, or adding a `MD_DEFAULTS_VERSION_LOWEST_SUPPORTED`?

puiterwijk commented on modulemd/v2/modulemd-defaults.c:
```diff
@@ -0,0 +1,309 @@
+      /* Already at this version, just copy it and return that */
+      return modulemd_defaults_copy (self);
+    }
+
```
In this case, an error isn't set.

puiterwijk commented on modulemd/v2/modulemd-defaults.c:
```diff
@@ -0,0 +1,309 @@
+      modulemd_defaults_set_module_name (self, g_value_get_string (value));
+      break;
+
+    case PROP_MDVERSION:
```
Maybe instead return an explicit error that while `MDVERSION` is valid, it can't be set?

sgallagher replied:
That's exactly what this does. It falls through to the default clause which does `G_OBJECT_WARN_INVALID_PROPERTY_ID` which means "this property isn't valid for this call".

puiterwijk replied:
Right, I meant to suggest not using the fallthrough and returning an explicit "This attribute is not supposed to be modified" error message.
But if you think the standard error suffices, fine with me.

puiterwijk commented on modulemd/v2/modulemd-defaults.c:
```diff
@@ -0,0 +1,309 @@
+    "Metadata Version",
+    "The metadata version of this Defaults object. Defaults to the "
+    "highest-supported version if not specified at object construction. "
+    "Cannot be changed after object construction except through the upgrade() "
```
Note that the `upgrade()` routine doesn't modify this on the object it receives, but creates a new one.
Thta makes this phrasing sound incorrect to me.

sgallagher replied:
You're right. I'll rephrase.

puiterwijk commented on modulemd/v2/tests/ModulemdTests/defaults.py:
```diff
@@ -0,0 +1,137 @@
+            with self.expect_signal():
+                defs = Modulemd.Defaults.new(
+                    Modulemd.DefaultsVersionEnum.LATEST + 1, 'foo')
+            assert False
```
Perhaps instead of this `try:` block, use `self.assertRaises`?
For an example, see the `Profile` patch: https://github.com/fedora-modularity/libmodulemd/pull/113/files#diff-3a8af7909cf33247e7740e17b3bcfb58R60.

sgallagher replied:
Ah, I didn't see that in the other patch. Handy!

sgallagher replied:
Replaced all of these cases with self.assertRaisesRegex() so I can also verify that the failure message is correct.

puiterwijk commented on modulemd/v2/include/modulemd-2.0/private/test-utils.h:
```diff
@@ -15,9 +15,30 @@
+ * Since: 2.0
+ */
+void
+modulemd_test_signal_handler (int sig_num);
```
Given how often we try to run some code to check that it emits a signal, maybe it's an idea to instead make a macro that takes the signal and a function to run and sets up and asserts the signal handler?
Kind of like the `expect_signal()` in the python `base.py`?

sgallagher replied:
Yeah, I was thinking the same thing. I'll look into that in a follow-up patch.

puiterwijk commented:
CI test failure looks like just rawhide failure.

sgallagher commented:
Modified the description of the `mdversion` property and replaced all of the `TypeError` checks in the python tests with `assertRaisesRegex()`. As before, I pushed this as a fixup commit so the differences are easier to spot, but if this is approved, I'll squash them together before merging.

sgallagher commented:
Also pushed another minor fix to have the `copy()` test verify that the `module_name` was also duplicated correctly.

sgallagher commented:
Rebased atop the Profile patch.
